File: Final Project/code/algorithms/hmc.py
import numpy as np
import physics
from algorithms.base import Base_sim
from numpy.random import Generator, default_rng
from typing import Optional, Callable
import physics.actions as actions
import algorithms.integrators as integrators
from physics.actions import BoundaryCondition

class HMCSim(Base_sim):
    """class to run the simulation with the HMC (Hybrid Monte Carlo) algorithm

    Args:
        Base_sim (abstract base class): HMCSim inherits from an abstract base class, which forces a common structure to other algorithms. 
    """
    def __init__(self, initial_phi: np.array, mass2: float, lamb: float, width: float, rng: Generator, integrator: Callable, bc: BoundaryCondition = BoundaryCondition.PBC):
        """initialize a HMC simulation

        Args:
            initial_phi (np.array): initial configuration
            mass2 (float): mass parameter, can be any real number (also negative)
            lamb (float): coupling strength of phi^4 theory, must be positive
            width (float): parameter to tune the acceptance rate, used to sample the configuration space
            rng (Generator): random number generator
            integrator (Callable): which integrator function to use
            bc (BoundaryCondition, optional): What kind of boundary conditions to use (period = PBC, aperiodic = APBC). Defaults to BoundaryCondition.PBC.
        """
        super().__init__()
        self.phi = initial_phi.copy()
        self.mass2 = float(mass2)
        self.lamb = float(lamb)
        self.eps = float(width)
        self.rng = np.random.Generator(np.random.PCG64(42))
        self.accepted_history = []
        self.integrator = integrator
        self.steps = 100
        self.bc = bc
        self.grad_action = actions.gradient_phi4_action
        self.grad_kwargs = dict(mass2=self.mass2, lamb=self.lamb, bc = self.bc)
        self.hamiltonian_diffs = []

    def update(self) -> list:
        """performs a single HMC algorithm step
        """
        pi0 = self.rng.normal(size=self.phi.shape)
        proposed_phi = self.phi.copy()
        proposed_pi = pi0.copy()
        diff_list = []

        for _ in range(self.steps): 
            proposed_phi, proposed_pi = self.integrator(proposed_phi, proposed_pi, self.eps, grad_action=self.grad_action, grad_kwargs=self.grad_kwargs)
        
        H_old = actions.hamiltonian(self.phi, pi0, mass2=self.mass2, lamb=self.lamb, bc = self.bc)
        H_new = actions.hamiltonian(proposed_phi, proposed_pi, mass2=self.mass2, lamb=self.lamb, bc = self.bc)
        p_accept = min(1, np.exp(H_old - H_new))
        diff_list.append(H_old - H_new)
        if self.rng.random() <= p_accept: 
            self.phi = proposed_phi
            self.accepted_history.append(1)
        else: 
            self.accepted_history.append(0)

        delta_H = H_old - H_new
        self.hamiltonian_diffs.append(delta_H)

def run_hmc(phi0: np.array, mass2: float, lamb: float, width: float, n_steps: int, rng: Optional[Generator] = None, logger: Optional[Callable] = None, progress: bool = False, integrator: Callable = integrators.leapfrog, return_diff: bool = False) -> HMCSim:
    """function to run the HMC simulation, a HMCSim object is initialized and the simulation is run with the desired parameters. 

    Args:
        phi0 (np.array): initial configuration
        mass2 (float): mass parameter, can be any real number (also negative)
        lamb (float): coupling strength of phi^4 theory, must be positive
        width (float): parameter to tune the acceptance rate, used to sample the configuration space
        n_steps (int): number of steps to run the simulation
        rng (Optional[Generator], optional): random number generator. Defaults to None.
        logger (Optional[Callable], optional): Logger function to use. Defaults to None.
        progress (bool, optional): Whether to display a progress bar or not. Defaults to False.
        integrator (Callable, optional): Which integrator to use. Defaults to integrators.leapfrog.
        return_diff (bool, optional): Whether to return the hamiltonian difference or not, used to check for energy conservation.

    Returns:
        HMCSim: A HMCSim object, the accepted history is returned
    """
    rng = rng or default_rng()
    sim = HMCSim(phi0, mass2=mass2, lamb=lamb, width=width, rng=rng, integrator=integrator)
    sim.run_sim(n_steps, logger=logger, progress_bar=True)
    if return_diff:
        return np.array(sim.hamiltonian_diffs)
    else:
        return sim.accepted_history

File: Final Project/code/rust/src/hmc.rs
use rand::Rng;

pub struct HCMSim<R, I>
where 
    R: Rng, 
    I: Integrator, 
{
    pub phi: Vec<f64>, //no longer a np.array

    pub mass2: f64, 
    pub lamb: f64, 
    pub eps: f64, 

    pub rng: R, 

    pub accepted_history: Vec<u8>, 
    pub integrator: I, //create your own type
    pub steps: usize, 
    pub bc: BoundaryCondition, 
}

impl<R, I> BaseSim for HCMSim<R, I>
where
    R: Rng, 
    I: Integrator, 
{
    fn update(&mut self) {
        let mut pi: Vec<f64> = self.phi.iter().map(|_| self.rng.sample(rand::distributions::StandardNormal)).collect();  // sample pi (initial momentum) from standard normal distribution (var = 1, mean = 0)

        let mut proposed_phi = self.phi.clone(); 
        let mut proposed_pi = pi.clone(); 

        for _ in 0..self.steps { //call the integrator (leapfrog or omelayan2 or maybe later some other)
            self.integrator.step(
                &mut proposed_phi,
                &mut proposed_pi,
                self.eps,
                self.mass2,
                self.lamb,
                self.bc,
            );
        }

        let h_old = hamiltonian(&self.phi, &pi, self.mass2, self.lamb, self.bc);
        let h_new = hamiltonian(&proposed_phi, &proposed_pi, self.mass2, self.lamb, self.bc);

        let p_accept = (h_old - h_new).exp().min(1.0);

        if self.rng.gen::<f64>() <= p_accept {
            self.phi = proposed_phi: 
            self.accepted_history.push(1);
        }else{
            self.accepted_history.push(0); 
        }
    }
    fn get_steps(&self) -> usize{
        self.steps
    }
}

//free run simulation function 
pub fn run_hmc<R, I>(
    phi0: Vec<f64>,
    mass2: f64,
    lamb: f64,
    width: f64,
    n_steps: usize,
    rng: R,
    integrator: I,
    logger: Option<impl FnMut(&HMCSim<R, I>)>,
) -> Vec<u8>
where
    R: Rng,
    I: Integrator,
{
    let mut sim = HMCSim {
        phi: phi0,
        mass2,
        lamb,
        eps: width,
        rng,
        accepted_history: Vec::new(),
        integrator,
        steps: 100,
        bc: BoundaryCondition::PBC,
    };

    // Run simulation
    sim.run_sim(n_steps, logger);

    // return acceptance history
    sim.accepted_history
}
